### PR TITLE
Retry policy decision: use current or next host

### PR DESCRIPTION
**Pull request by jorgebay**
Includes `RetryPolicy` jsdoc improvements (NODEJS-225) and new `#onRequestError()` method (NODEJS-251).

### CODE REVIEW

File: lib/policies/retry.js [hunk 1]
```diff
@@ -56,22 +65,30 @@ RetryPolicy.prototype.onWriteTimeout = function (requestInfo, consistency, recei
 };
 
 /**
- * @returns {{decision: Number}}
+ * Returns a {@link DecisionInfo} to retry the request with the given [consistency]{@link module:types~consistencies}.
+ * @param {Number} [consistency] When specified, it retries the request with the given consistency.
+ * @param {Boolean} [useCurrentHost] When specified, determines if the retry should be made using the same coordinator.
```
Comment (tolbertam) [inline lib/policies/retry.js]: Would be good to indicate the default is to not try the next host.

File: lib/request-handler.js [hunk 2]
```diff
@@ -323,24 +323,35 @@ RequestHandler.prototype.handleError = function (err, callback) {
...
   }
-  if (decisionInfo && decisionInfo.decision === retry.RetryPolicy.retryDecision.retry) {
-    return this.retry(callback);
+  if (decisionInfo) {
+    if (decisionInfo.decision === retry.RetryPolicy.retryDecision.retry) {
+      return this.retry(decisionInfo.consistency, decisionInfo.useCurrentHost, callback);
+    }
+    if (decisionInfo.decision === retry.RetryPolicy.retryDecision.ignore) {
+      //Return an empty response
+      return callback(
```
Comment (tolbertam) [inline lib/request-handler.js]: Was there not logic for ignore before, guess never noticed :)

File: lib/request-handler.js [hunk 3]
```diff
@@ -351,12 +362,25 @@ RequestHandler.prototype.handleError = function (err, callback) {
...
+  }
   if (this.retryHandler) {
+    //custom retry handler (not a QueryRequest / ExecuteRequest / BatchRequest)
     return this.retryHandler(callback);
   }
+  if (useCurrentHost !== false) {
+    // retry on the current host by default
+    return this.sendOnConnection(this.request, this.requestOptions, callback);
+  }
   this.send(this.request, this.requestOptions, callback);
```
Comment (tolbertam) [inline lib/request-handler.js]: Shouldn't this `this.send` not be called if useCurrentHost is true?

**Comment by tolbertam** [comment]
To be consistent with the java driver, should we update RetryPolicy to retry once on next host on unavailable by default?  ([java driver code that does this](https://github.com/datastax/java-driver/blob/3.0.1/driver-core/src/main/java/com/datastax/driver/core/policies/DefaultRetryPolicy.java#L108-L113)).  This was done as part of [JAVA-709](https://datastax-oss.atlassian.net/browse/JAVA-709) to address the possible case where a node gets partitioned from the rest of the cluster, but the client can still talk to it (seems unlikely, but it can happen).

**Comment by jorgebay** [comment]
> Should we update RetryPolicy to retry once on next host on unavailable by default?

I wasn't aware of that change, I'll fix it!

Also, I think that it makes sense to include [NODEJS-251](https://datastax-oss.atlassian.net/browse/NODEJS-251) in this patch, to group all the upcoming changes related to the `RetryPolicy`. What do you think?

**Comment by tolbertam** [comment]
> Also, I think that it makes sense to include NODEJS-251 in this patch, to group all the upcoming changes related to the RetryPolicy. What do you think?

I agree, this is a good place to put that. 👍

File: lib/request-handler.js (hunk 3 above)
Comment (jorgebay) [inline lib/request-handler.js]: I'm not sure if I understand the question, `#sendOnConnection ()` uses the current connection and `#send()` uses the next host in the query plan.

Comment (tolbertam) [inline lib/request-handler.js]: Ah, when I made the comment I must not have noticed the `return` in the sendOnConnection call, in that case this is fine.

File: lib/request-handler.js [hunk 4]
```diff
@@ -295,71 +300,109 @@ RequestHandler.prototype.prepareOnConnection = function (queries, callbacksArray
...
+/**
+ * @returns {{decision, useCurrentHost, consistency}}
+ */
+RequestHandler.prototype.getDecision = function (err) {
+  var requestInfo = {
+    request: this.request,
+    handler: this,
+    nbRetry: this.retryCount
+  };
+  //TODO: isServerUnhealthy is going to be called isSocketError as part of NODEJS-106 and should not call setDown()
```
Comment (jorgebay) [inline lib/request-handler.js]: We should revisit this pull request once NODEJS-106 is merged, as it affects how socket error flag is named and the behaviour when there is a socket error.

**Comment by jorgebay** [comment]
- Included `onRequestError()` method in the retry policy.
- Added the flag `wasRequestWritten` when we suspect that it could have been written in the wire (and mutation could be applied). We retry on next host if we know for a fact that the request was not written and rely on the retry policy otherwise. 
- Rebased against master (after #145 was merged).
- The retry policy behaviour is supposed to be the same as the `DefaultRetryPolicy` in the java driver 3.0+.

**Comment by tolbertam** [comment]
I wrote some tests to validate this using [stubbed cassandra](http://www.scassandra.org/).  Since the test requires standing stubbed cassandra clusters I put it on a [separate branch](https://github.com/tolbertam/nodejs-driver/tree/NODEJS-105-stubbed) with the eventual goal to integrate it into our test automation.  A run down of what is validated:

```
  RetryPolicy
    #onUnavailable()
      ✓ should retry on the next host the first time (50ms)
      ✓ should rethrow the following times
    #onRequestError()
      ✓ should retry on the next host until successful (2033ms)
      ✓ should rethrow when retryOnTimeout is false (1019ms)
      ✓ should retry on the next host until successful if connection closed on request
      ✓ should retry on the next host until successful if server error
```

Some of these tests are modeled after the unit tests (just at an integration level).  Would appreciate any early feedback on that branch as I intend to build up a stubbed suite for this and also #145 among other things.

The only thing I haven't been able to figure out how to exercise is when wasRequestWritten != true (which you have a unit test for via `should retry when there was a socket error and mutation was not applied`)

All that being said, this looks good to me. 👍

**Comment by jorgebay** [comment]
Big +1 to tests with stubbed cassandra, I've commented on some minor improvements on the commit.

**Pull request merged**